patternmatch.go

```diff
@@ -93,9 +106,139 @@ func patternMatchOperands(pattern, expr Expr, varCache map[VarName]Expr) bool {
 	for ix := 1; ix <= NumberOfOperands(pattern); ix++ {
 		patternOp := Operand(pattern, ix)
 		exprOp := Operand(expr, ix)
-		if !patternMatch(patternOp, exprOp, varCache) {
+		if !patternMatchOld(patternOp, exprOp, varCache) {
 			return false
 		}
 	}
 	return true
 }
+
+func patternMatch(expr, pattern Expr, bindings Binding) bool {
+	switch p := pattern.(type) {
+	case undefined:
+		if _, ok := expr.(undefined); ok {
+			return true
+		}
+		return false
+
+	case constant:
+		if c, ok := expr.(constant); ok {
+			return c.Value == p.Value
+		}
+		return false
+
+	case variable:
+		// If the varaible in the pattern is not
+		// a pattern variable, it means that we are
+		// matching expr against a binding. This is a
+		// base case and we check for equality
+		if !p.isPattern {
+			return Equal(p, expr)
+		}
+
+		// Extracts binding to pattern variable if it exists
+		// and gets if expression is a variable
+		boundExpr, bindingExists := bindings[p.Name]
+
+		// If no expression is bound to this variable
+		// we bound the current expression to it and return
+		// true
+		if !bindingExists {
+			bindings[p.Name] = expr
+			return true
+		}
+
+		// If bound expression exist, we recursively
+		// match expr against it
+		return patternMatch(expr, boundExpr, bindings)
+
+	case constrainedVariable:
+		// If the varaible in the pattern is not
+		// a pattern variable, it means that we are
+		// matching expr against a binding. This is a
+		// base case and we check for equality
+		if !p.isPattern {
+			return Equal(p, expr)
+		}
+
+		// Extracts binding to pattern variable if it exists
+		// and gets if expression is a variable
+		boundExpr, bindingExists := bindings[p.Name]
+
+		// If no expression is bound to this variable
+		// we bound the current expression to it and return
+		// true
+		if !bindingExists && p.Constraint(expr) {
+			bindings[p.Name] = expr
+			return true
+		} else if !p.Constraint(expr) {
+			return false
+		}
+
+		// If bound expression exist, we recursively
+		// match expr against it
+		return patternMatch(expr, boundExpr, bindings)
+
+	case add:
+		if a, ok := expr.(add); ok {
+			// Checks that expr and pattern contains the
+			// same number of operands
+			nOpPattern := NumberOfOperands(p)
+			nOpExpr := NumberOfOperands(a)
+			if nOpPattern != nOpExpr {
```

Comment:
```suggestion
			if  NumberOfOperands(p)!= NumberOfOperands(a) {
```
You also reuse the function below on expr (that should give the same result as NumberOfOperands(a), right?) so I don't see the purpose of defining a variable here

Reply:
I see your point. However, the `NumberOfOperands` is a recursive call which I do not want to do too often. In this specific case it is probably not going to be problematic, but just to stick to good practice, I suggest reusing `nOpExpr`. What do you think?

Reply:
it makes sense, what about the other call few lines below, maybe that one can be changed with nOpExpr